cli/src/commands/rebase.rs

```diff
@@ -509,69 +457,285 @@ fn extract_commits(
+    let mut to_visit = dag_walk::topo_order_reverse(
+        to_visit_commits.keys().cloned().collect_vec(),
+        |commit_id| commit_id.clone(),
+        |commit_id| -> Vec<CommitId> {
+            visited.insert(commit_id.clone());
+            to_visit_commits_new_parents
+                .get(commit_id)
+                .cloned()
+                .unwrap()
+                .iter()
+                // Only add parents which are in the set to be visited and have not already been
+                // visited.
+                .filter(|&id| to_visit_commits.contains_key(id) && !visited.contains(id))
+                .cloned()
+                .collect()
         },
-    )?;
+    );
 
-    Ok((target_roots, num_rebased_descendants))
+    let mut skipped_commits: Vec<Commit> = vec![];
+    let mut num_rebased = 0;
+    let mut num_rebased_descendants = 0;
+
+    // Rebase each commit onto its new parents in the reverse topological order
+    // computed above.
+    // TODO(ilyagr): Consider making it possible for descendants of the target set
+    // to become emptied, like --skip-empty. This would require writing careful
+    // tests.
+    while let Some(old_commit_id) = to_visit.pop() {
+        let old_commit = to_visit_commits.get(&old_commit_id).unwrap();
+        let parent_ids = to_visit_commits_new_parents
+            .get(&old_commit_id)
+            .cloned()
+            .unwrap();
+        let new_parent_ids = tx.mut_repo().new_parents(parent_ids);
+        let rewriter = CommitRewriter::new(tx.mut_repo(), old_commit.clone(), new_parent_ids);
+        if rewriter.parents_changed() {
+            let builder = rewriter.rebase(settings)?;
+            builder.write()?;
+            if target_commit_ids.contains(&old_commit_id) {
+                num_rebased += 1;
+            } else {
+                num_rebased_descendants += 1;
+            }
+        } else {
+            skipped_commits.push(old_commit.clone());
+        }
+    }
+    tx.mut_repo().update_rewritten_references(settings)?;
+
+    if let Some(mut fmt) = ui.status_formatter() {
+        if !skipped_commits.is_empty() {
+            log_skipped_rebase_commits_message(
+                fmt.as_mut(),
+                tx.base_workspace_helper(),
+                skipped_commits.iter(),
+            )?;
+        }
+        if num_rebased > 0 {
+            writeln!(fmt, "Rebased {num_rebased} commits onto destination")?;
+        }
+        // TODO: depending on the destination, this description isn't really accurate.
+        // Should we just say rebased X other commits?
+        if num_rebased_descendants > 0 {
+            writeln!(
+                fmt,
+                "Rebased {num_rebased_descendants} descendant commits onto parents of rebased \
+                 commits"
+            )?;
+        }
+    }
+
+    if tx.mut_repo().has_changes() {
+        tx.finish(ui, tx_description)
+    } else {
+        Ok(()) // Do not print "Nothing changed."
```

Comment:
I believe calling `tx.finish` when the repository is unchanged will display the message, there are other such checks around the codebase.

Reply:
This would be possible if all the commits were already in the right location, and nothing was actually rewritten (which is what the skipped rebase message shows).

Reply:
Printing "Nothing changed." in that case seems somewhat useful to me. At least it doesn't seem so confusing that we want to avoid it.

Reply:
Updated to just print the message if nothing changed.